File: project-{topic}-optimized/backend/src/main/java/com/alx/chat/repository/UserRepository.java
```java
package com.alx.chat.repository;

import com.alx.chat.entity.User;
import org.springframework.data.jpa.repository.JpaRepository;
import org.springframework.stereotype.Repository;

import java.util.Optional;

/**
 * Repository for managing {@link User} entities.
 */
@Repository
public interface UserRepository extends JpaRepository<User, Long> {
    /**
     * Finds a user by their username.
     * @param username The username to search for.
     * @return An Optional containing the User if found.
     */
    Optional<User> findByUsername(String username);

    /**
     * Finds a user by their email.
     * @param email The email to search for.
     * @return An Optional containing the User if found.
     */
    Optional<User> findByEmail(String email);

    /**
     * Checks if a user exists with the given username.
     * @param username The username to check.
     * @return True if a user with the username exists, false otherwise.
     */
    boolean existsByUsername(String username);

    /**
     * Checks if a user exists with the given email.
     * @param email The email to check.
     * @return True if a user with the email exists, false otherwise.
     */
    boolean existsByEmail(String email);
}
```

File: project-{topic}-advanced/backend/src/repositories/User.repository.ts
```typescript
import { Repository } from 'typeorm';
import { User } from '../entities/User.entity';

export class UserRepository extends Repository<User> {
  // Custom methods can be added here if needed, e.g., findByEmail
  async findByEmail(email: string): Promise<User | null> {
    return this.findOne({ where: { email } });
  }

  async findByUsername(username: string): Promise<User | null> {
    return this.findOne({ where: { username } });
  }

  async findById(id: string): Promise<User | null> {
    return this.findOne({ where: { id } });
  }
}
```